### ISSUE TITLE
Tensor dimension becomes None after using gather or boolean_mask in Tensorflow 2.0

### ISSUE THREAD

**Issue by kupferb** [tensorflow/tensorflow#53220]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): Yes
- Windows 10 Home (19043.1348):
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): 2.7.0
- Python version: 3.8.10
- CUDA/cuDNN version: 11.2
- GPU model and memory: GTX 1080 Ti (11 GB)

**Describe the current behavior**

For some reason I am getting different tensor dimensions when using `gather` in TF 2:

1. The first dimension becomes `None` when I use tensor as an index vector 
2. The first dimension becomes `len(indices)` (as it should) where 'indices' are regular Python list

**This happens only in eager mode (e.g., inside a custom loss function)**

(Same happens when using `boolean_mask`)

**Standalone code to reproduce the issue**

    import tensorflow as tf
    from tensorflow import keras
    from tensorflow.keras import Model
    from tensorflow.keras.layers import Input, Dense, Reshape
    from tensorflow.keras.datasets import mnist
 
    def cutsom_gan_loss_env(model):
       def custom_loss(y_true,y_pred):

        ff = tf.where([True, True, False , False])[:, 0]
        with tf.GradientTape(persistent=True) as tape:
             tf.print(tf.gather(y_true, [0, 1], axis=0).shape) #prints 2,28,28
             tf.print(tf.gather(y_true, ff, axis=0).shape)#prints None,28,28
             tape.watch(y_true)
             yy = model(y_true)
             d_yy = tape.gradient(yy,y_true)
             des_loss = tf.reduce_mean(d_yy)

        return des_loss

    return custom_loss

    def main_():
       n_hidden_units = 5
       num_lay = 3
       kernel_init = keras.initializers.RandomUniform(-0.1, 0.1)
       (x_train, y_train), _ = mnist.load_data()
       x_train = tf.cast(x_train,tf.float32)/255.
       inputs = Input(x_train.shape[1:])
       x = Dense(n_hidden_units,kernel_initializer=kernel_init,  activation='sigmoid' )(inputs)
       for _ in range(num_lay):
           x = Dense(n_hidden_units,kernel_initializer=kernel_init, activation='sigmoid', )(x)

       outputs =Reshape(x_train.shape[1:])(Dense(x_train.shape[1], kernel_initializer=kernel_init, activation='softmax')(x))
       model = Model(inputs=inputs, outputs=outputs)
       model.summary()
       optimizer1 = keras.optimizers.Adam(beta_1=0.9, beta_2=0.999, epsilon=None, decay=0.0, amsgrad=True)
       model.compile(loss=cutsom_gan_loss_env(model), optimizer=optimizer1, metrics=None)
       model.fit(x_train,  x_train , batch_size=1000, epochs=1, shuffle=False)

    if __name__=='__main__':
        main_()

**Comment by sushreebarsa** [comment]
@kupferb 
Please post this issue on [keras-team/keras repo.](https://github.com/keras-team/keras/issues)
To know more see;
[https://discuss.tensorflow.org/t/keras-project-moved-to-new-repository-in-https-github-com-keras-team-keras/1999](https://discuss.tensorflow.org/t/keras-project-moved-to-new-repository-in-https-github-com-keras-team-keras/1999)
Thank you!
